### ISSUE TITLE
Documentation Request: How to mock useQuery refetch method

### ISSUE THREAD

**Issue by coler-j** [apollographql/apollo-client#7286]
This is a question that appears to have been repeatedly asked, but due to archiving repos and package evolution there is no documented answer. Spectrum discussion is ineffective at surfacing and maintaining knowledge of a solution; this is something that should be added to the documentation or at the least kept as a historic record within github issues.

# Question

> How to mock useQuery refetch method? How to mock refetch with different input variables?

Passed unanswered / closed questions:

- https://spectrum.chat/apollo/testing/is-it-possible-to-mock-refetch~742a6083-b8a0-4bdf-9a6c-d2b512d16c41
- https://stackoverflow.com/questions/58981274/is-there-a-way-to-mock-refetch-in-mockedprovider-apollo-client
- https://github.com/apollographql/react-apollo/issues/3793
- https://github.com/apollographql/react-apollo/issues/2783
- https://github.com/apollographql/react-apollo/issues/3190
- (my own spectrum question) https://spectrum.chat/apollo/apollo-client/how-to-test-query-refetch~6f7b5342-2375-4e90-b846-6a0225e82d11

Ideally the solution would be updated to these documents: https://www.apollographql.com/docs/react/development-testing/testing/#mockedprovider

**Comment by coler-j** [comment]
As a workaround, for the few tests that utilize `refetch` I had to physically mock the `useQuery` function and provide mock functions for the return of `refetch`; for our custom hook (where an overridden `useQuery` hook is exported as default), it looked something like this:

```
import * as useQueryModule from '~/hooks/useQuery';

describe('test the thing', () => {
    let useQuerySpy;

    beforeEach(() => {
        // Spy on the `useQuery` function so we can optionally provide alternate behaviour.
        useQuerySpy = jest.spyOn(useQueryModule, 'default');
    })

    afterEach(() => {
      // Restore any mocked behaviour
      useQuerySpy.mockRestore();
    });

    it('does a thing', async () => {
        const refetchedApolloResponse = buildResponse(refetchData) // some function to build the shape / data of apollo response
        const initialApolloResponse = buildResponse(initialData) // some function to build the shape / data of apollo response
     
       const mockRefetch = jest.fn().mockResolvedValue({ data: refetchedApolloResponse });
       useQuerySpy.mockReturnValue({ data: initialApolloResponse, refetch: mockRefetch });

      // Assert stuff
    }
})
```

**Comment by dylanwulf** [comment]
I prefer to use this approach for mocking GraphQL requests: https://www.freecodecamp.org/news/a-new-approach-to-mocking-graphql-data-1ef49de3d491/

It eliminates the need for MockedProvider, and refetch works without any extra hassle

**Comment by jbanulso** [comment]
I explain this in: https://stackoverflow.com/questions/58981274/is-there-a-way-to-mock-refetch-in-mockedprovider-apollo-client/65076775#65076775:

TL;DR:
```
let queryCalled = false;
const refetchMock = {
  request: {
    query: YOUR_QUERY
  },
  newData: () => {
    if (queryCalled) {
      return { 
        data: {
          // your refetched data
        }
      }; 
    } else {
      queryCalled = true;
      return { 
        data: {
          // your first fetched data
        } 
      }; 
    }
  }
};

**Comment by hwillson** [comment]
It doesn't sound like there is an outstanding issue here, so closing. Thanks!
